lib/spack/spack/package.py

```diff
@@ -1274,8 +1274,14 @@ def compiler(self):
         """Get the spack.compiler.Compiler object used to build this package"""
         if not self.spec.concrete:
             raise ValueError("Can only get a compiler for a concrete package.")
-        return spack.compilers.compiler_for_spec(self.spec.compiler,
-                                                 self.spec.architecture)
+
+        if not hasattr(self, '_compiler'):
+            self._compiler = spack.compilers.compiler_for_spec(
+                self.spec.compiler,
+                self.spec.architecture
+            )
+
+        return self._compiler
```

Comment:
Could you wrap `def compiler` in `@llnl.util.memoized` for the same effect but with less new code?

Reply:
I never tried with a property but it should work. The order in case is:
```python
@property
@llnl.util.lang.memoized
def compiler(self):
    ....
```

Reply:
Meh, it seems we're hitting this https://github.com/python/mypy/issues/1362 This is the first time we decorate a property, but:
1. It's legit and works on Python 2
2. The error detected is from an open issue that is stale since 2016?

So it feels somehow not an improvement to fix this with:
```python
@property  # type: ignore
@llnl.util.lang.memoized
def compiler(self):
```
Can't we disable this check from mypy? @trws

Reply:
We cannot, other than with the ignore comment or ignoring the file.  There just isn't an option to skip this error for whatever reason.

Reply:
We *could* add an internal function or method that is not a property with the memoized annotation that the property calls though.  That would avoid the entire problem.

Reply:
> We could add an internal function or method that is not a property with the memoized annotation that the property calls though. That would avoid the entire problem.

It doesn't seem a better solution. We are splitting a function into two just to work around an issue in a linter. Not a major problem, but it was somehow unexpected to see that nobody fixed it since 2016.